### DISCUSSION TITLE
Can a Custom Node also be a 'Group' node?

### DISCUSSION THREAD

**Discussion by ChuckBuckster** [xyflow/xyflow#2592] [Q&A]
I have created a custom node, which uses the `react-bootstrap` progress bar component in order to represent the progress of a pipeline.  

```javascript
const CustomFlowComponent = ({data}) => {

    useEffect(() => {
        console.log(data);
    }, [])

    return (
        <Grid container display={'flex'} alignContent={'center'} border={1} borderRadius={"5px"} width={200}
              sx={{backgroundColor: '#1e1e1e'}}>
            <Handle type="target" position={Position.Left}/>
            <Grid item xs={1}/>
            <Grid item xs={10} paddingY={1}>
                <Typography fontWeight={"bold"} color={"orange"}>
                    {data.label}
                </Typography>
            </Grid>
            <Grid item xs={1}/>
            <Grid item xs={12} sx={{height: "5rem"}}>
                {/* hardcoded test data for now*/}
                <ProgressBar>
                    <ProgressBar variant="success" now={50} key={1}/>
                    <ProgressBar animated variant="warning" now={40} key={2}/>
                    <ProgressBar variant="danger" now={10} key={3}/>
                </ProgressBar>
            </Grid>
            <Handle type="source" position={Position.Right}/>
        </Grid>
    );
}

export default CustomFlowComponent;
```

In my main component, I have some calculation function(s) which generate the nodes based on a large set of Object data coming in:

```javascript
        pipelineOrder.map(pipelineGroupName => {
            let currentNode = {};
            currentNode.id = pipelineGroupName;
            currentNode.type = 'customNode';
            const currentPosition = {x: previousNodePosition.x + 300, y: previousNodePosition.y + 0}
            currentNode.position = currentPosition;
            previousNodePosition = currentPosition;
            const groupData = pipelineGroupData[pipelineGroupName];
            // get all pipelines from specified group
            Object.keys(groupData).map((pipelineName, pipelineNameIndex) => {
                currentNode.data = {label: pipelineGroupName, pipelineData: groupData[pipelineName]};
            })
            calculatedNodes = [...calculatedNodes, currentNode];
        })
```

Higher up, as per the doc example, I am using this to reference my use of my custom node:
```javascript
const nodeTypes = useMemo(() => ({customNode: CustomFlowComponent}), []);
```

And it gets render in the return function here:
```javascript
    return (
        <Grid container sx={{height: 1000}}>
            <ReactFlow nodes={nodes} edges={edges} nodeTypes={nodeTypes} onNodesChange={onNodesChange}
                       onEdgesChange={onEdgesChange}>
                <Background/>
                <Controls/>
            </ReactFlow>
        </Grid>
    );
```

So, when you would normally want to denote a node as `type: 'group'`, you can't. Or can you..?

Is there a way to use a custom node **and** also set it as `type: 'group'`, so I can add child-nodes within it?

**Comment by Schekhovtsov** [comment]
Hi! I have the same question.

**Comment by moklick** [comment]
Every node can act as a parent. When another uses the id for `parentNode`, it gets a parent.

**Reply by tarun-2002** [reply]
But How can we implement this with custom nodes. Which have drag and drop functionality.

**Reply by bcakmakoglu** [reply]
You have to be more specific about your question here, otherwise it's hard to help you out 😄 
What isn't working for you? What have you tried? Can you give us an idea through some code a reproduction or something similar?

**Comment by HiMo-LGCNS** [comment]
`
import React, { forwardRef, HTMLAttributes, ReactNode } from "react";
import { NodeProps, Panel, PanelPosition } from "@xyflow/react";
import { BaseNode } from "@/registry/components/base-node";
import { cn } from "@/lib/utils";

/* GROUP NODE Label ------------------------------------------------------- */

export type GroupNodeLabelProps = HTMLAttributes<HTMLDivElement>;

export const GroupNodeLabel = forwardRef<HTMLDivElement, GroupNodeLabelProps>(
  ({ children, className, ...props }, ref) => {
    return (
      <div ref={ref} className="h-full w-full" {...props}>
        <div
          className={cn(
            "w-fit bg-gray-200 bg-secondary p-2 text-xs text-card-foreground",
            className,
          )}
        >
          {children}
        </div>
      </div>
    );
  },
);

GroupNodeLabel.displayName = "GroupNodeLabel";

export type GroupNodeProps = Partial<NodeProps> & {
  label?: ReactNode;
  position?: PanelPosition;
};

/* GROUP NODE -------------------------------------------------------------- */

export const GroupNode = forwardRef<HTMLDivElement, GroupNodeProps>(
  ({ selected, label, position, ...props }, ref) => {
    const getLabelClassName = (position?: PanelPosition) => {
      switch (position) {
        case "top-left":
          return "rounded-br-sm";
        case "top-center":
          return "rounded-b-sm";
        case "top-right":
          return "rounded-bl-sm";
        case "bottom-left":
          return "rounded-tr-sm";
        case "bottom-right":
          return "rounded-tl-sm";
        case "bottom-center":
          return "rounded-t-sm";
        default:
          return "rounded-br-sm";
      }
    };

    return (
      <BaseNode
        ref={ref}
        selected={selected}
        className="h-full overflow-hidden rounded-sm bg-white bg-opacity-50 p-0"
        {...props}
      >
        <Panel className={cn("m-0 p-0")} position={position}>
          {label && (
            <GroupNodeLabel className={getLabelClassName(position)}>
              {label}
            </GroupNodeLabel>
          )}
        </Panel>
      </BaseNode>
    );
  },
);

GroupNode.displayName = "GroupNode";

`


this is core group you can customize it
